src/fpnv/token-verifier.ts

```diff
@@ -0,0 +1,246 @@
+      );
+    }
+
+    const projectId = await this.ensureProjectId();
+    const decoded = await this.decodeAndVerify(jwtToken, projectId);
+    const decodedIdToken = decoded.payload as FpnvToken;
+    decodedIdToken.getPhoneNumber = () => decodedIdToken.sub;
+    return decodedIdToken;
+  }
+
+  private async ensureProjectId(): Promise<string> {
+    const projectId = await util.findProjectId(this.app);
+    if (!validator.isNonEmptyString(projectId)) {
+      throw new FirebasePnvError(
+        FpnvErrorCode.COMMON_ISSUE,
+        'Must initialize app with a cert credential or set your Firebase project ID as the ' +
+                `GOOGLE_CLOUD_PROJECT environment variable to call ${this.tokenInfo.verifyApiName}.`);
+    }
+    return projectId;
+  }
+
+  private async decodeAndVerify(
+    token: string,
+    projectId: string,
+  ): Promise<DecodedToken> {
+    const decodedToken = await this.safeDecode(token);
+    this.verifyContent(decodedToken, projectId);
+    await this.verifySignature(token);
+    return decodedToken;
+  }
+
+  private async safeDecode(jwtToken: string): Promise<DecodedToken> {
+    try {
+      return await decodeJwt(jwtToken);
+    } catch (err) {
+      if (err.code === JwtErrorCode.INVALID_ARGUMENT) {
+        const verifyJwtTokenDocsMessage = ` See ${this.tokenInfo.url} ` +
+                    `for details on how to retrieve ${this.shortNameArticle} ${this.tokenInfo.shortName}.`;
+        const errorMessage = `Decoding ${this.tokenInfo.jwtName} failed. Make sure you passed ` +
+                    `the entire string JWT which represents ${this.shortNameArticle} ` +
+                    `${this.tokenInfo.shortName}.` + verifyJwtTokenDocsMessage;
+        throw new FirebasePnvError(FpnvErrorCode.COMMON_ISSUE,
+          errorMessage);
+      }
+      throw new FirebasePnvError(FpnvErrorCode.COMMON_ISSUE, err.message);
+    }
+  }
+
+
+  private verifyContent(
+    fullDecodedToken: DecodedToken,
+    projectId: string | null,
+  ): void {
+    const header = fullDecodedToken && fullDecodedToken.header;
+    const payload = fullDecodedToken && fullDecodedToken.payload;
+
+    const projectIdMatchMessage = ` Make sure the ${this.tokenInfo.shortName} comes from the same ` +
+            'Firebase project as the service account used to authenticate this SDK.';
+    const verifyJwtTokenDocsMessage = ` See ${this.tokenInfo.url} ` +
+            `for details on how to retrieve ${this.shortNameArticle} ${this.tokenInfo.shortName}.`;
+
+    let errorMessage: string | undefined;
+
+    // JWT Header
+    if (typeof header.kid === 'undefined') {
+      errorMessage = `${this.tokenInfo.jwtName} has no "kid" claim.`;
+      errorMessage += verifyJwtTokenDocsMessage;
+    } else if (header.alg !== ALGORITHM_ES256) {
+      errorMessage = `${this.tokenInfo.jwtName} has incorrect algorithm. Expected "` + ALGORITHM_ES256 + '" but got ' +
+                '"' + header.alg + '".' + verifyJwtTokenDocsMessage;
```

Comment:
Let's use template strings consistently for these messages (you may still need to use `+` in order to split onto multiple lines)